### PR TITLE
Add type definitions for zEdit UPF and xelib

### PR DESCRIPTION
**Pull request by awlayton**
This adds type definitions for [zEdit's UPF](https://github.com/z-edit/zedit-unified-patching-framework) modules and [xelib](https://github.com/z-edit/xelib) (xelib is a dependency of UPF).

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Review by sheetalkamat — CHANGES_REQUESTED** [review]
https://github.com/z-edit/xelib#installation says its going to be on npm soon at z-edit/xelib so we would want to wait for that to happen before publishing conflicting definitions since in that case the location of definition isnt xelib but `z-edit__xelib` which you cant because the package is not yet there and we cant guarantee that till that actually happens

File: types/xelib/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,1629 @@
...
+     * Active plugins are returned as a list of filenames separated by \r\n.
+     */
+    GetActivePlugins(): string;
+    /**
+     * Loads plugin files in loadOrder.
+     * If smartLoad is set to true, master files required by files in loadOrder will be automatically loaded as necessary.
+     * Plugin loading is performed in a background thread, use GetLoaderStatus to track the loader and determine when it is done.
+     * @see GetLoaderStatus
+     */
+    LoadPlugins(loadOrder: string, smartLoad: boolean): void;
```
Comment (peterblazejewicz) [inline types/xelib/index.d.ts:474]:
```suggestion
    LoadPlugins(loadOrder: string, smartLoad?: boolean): void;
```
https://github.com/z-edit/xelib/blob/72c8dd97956098947b49fbdeadd0e32fd8ca0d38/src/js/setup.js#L80

File: types/xelib/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,1629 @@
...
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 3.4
+
+/**
+ * xelib instance
+ */
+export const wrapper: XELib;
+
+// Distinguish handles from regular numbers
+export type Handle = number & { __xelib_handle__: never };
```
Comment (peterblazejewicz) [inline types/xelib/index.d.ts:13]: is that required? The code from README would have to be update to:
```ts
xelib.GetElement(0 as Handle, 'Skyrim.esm\\00012E46');
```
up to you folks, though.

Comment (awlayton) [reply types/xelib/index.d.ts:13]: The library is actually a wrapper around a dll. The handles are only meaningful to the underlying code and a user of the library should not be passing anything that was not received from the library as a Handle.

There are cases where functions (e.g., `GetElement`) take `0` in place of a Handle. I tried to make all such functions accept `0` as well as `Handle` but it appears I missed that one.

I will get rid of that and just alias `Handle` to `number` if that is preferred.

**Comment by awlayton** [comment]
> https://github.com/z-edit/xelib#installation says its going to be on npm soon at z-edit/xelib so we would want to wait for that to happen before publishing conflicting definitions since in that case the location of definition isnt xelib but `z-edit__xelib` which you cant because the package is not yet there and we cant guarantee that till that actually happens

Yes I saw that, but I am not sure it will be very _soon_ (they put that there over a year ago). You think they are going to publish it as a scoped package rather than just `xelib`?

File: types/xelib/index.d.ts (hunk 2 above)
Comment (peterblazejewicz) [reply types/xelib/index.d.ts:13]: cool, that's perfectly fine with me. I've just pointed that in the readme-driven-development that nuance arise with that specific detail. 👍

**Review by peterblazejewicz — APPROVED** [review]
LGTM!

**Review by peterblazejewicz — APPROVED** [review]

**Comment by awlayton** [comment]
@sheetalkamat the command `npm i z-edit/xelib` is the _current_ command to install it, *not* how to install it once it is on npm.

I see no reason to not think it will be published as `xelib`. The name `zedit/xelib` is not even a valid name for npm is it? The package already has the name `xelib` in its package.json.

**Review by amcasey — APPROVED** [review]
It's unfortunate that they've muddied the water about what their eventual package will be called, but it doesn't seem like an update is imminent.

**Pull request merged**